service/dap/server_test.go

```diff
@@ -345,135 +345,163 @@ func TestSetBreakpoint(t *testing.T) {
+
+					client.StackTraceRequest(1, 3, 0)
+					stResp = client.ExpectStackTraceResponse(t)
+					expectFrames(t, stResp, 1015, 3, 6, 17)
+
+					client.StackTraceRequest(1, 6, 0)
+					stResp = client.ExpectStackTraceResponse(t)
+					expectFrames(t, stResp, 0, 0, 6, 0)
+
+					client.StackTraceRequest(1, 7, 0) // Out of bounds startFrame is capped at len
+					stResp = client.ExpectStackTraceResponse(t)
+					expectFrames(t, stResp, 0, 0, 6, 0)
+				},
+				// Stop at line 18
+				func() {
+					t.Helper()
+					// Frame ids get reset at each breakpoint.
+					client.StackTraceRequest(1, 0, 0)
+					stResp = client.ExpectStackTraceResponse(t)
+					expectFrames(t, stResp, 1000, 3, 3, 18)
+
+					client.StackTraceRequest(1, 0, -100) // Negative levels is treated as 0
+					stResp = client.ExpectStackTraceResponse(t)
+					expectFrames(t, stResp, 1003, 3, 3, 18)
+
+					client.StackTraceRequest(1, 0, 2)
+					stResp = client.ExpectStackTraceResponse(t)
+					expectFrames(t, stResp, 1006, 2, 3, 18)
+
+					client.StackTraceRequest(1, 0, 3)
+					stResp = client.ExpectStackTraceResponse(t)
+					expectFrames(t, stResp, 1009, 3, 3, 18)
+
+					client.StackTraceRequest(1, 0, 4) // Out of bounds levels is capped at len
+					stResp = client.ExpectStackTraceResponse(t)
+					expectFrames(t, stResp, 1012, 3, 3, 18)
+
+					client.StackTraceRequest(1, 1, 2)
+					stResp = client.ExpectStackTraceResponse(t)
+					expectFrames(t, stResp, 1016, 2, 3, -1) // Don't test for runtime line we don't control
+				}})
+	})
+}
 
-		client.ExpectTerminatedEvent(t)
-		client.DisconnectRequest()
-		client.ExpectDisconnectResponse(t)
+// Tests that 'stackTraceDepth' from LaunchRequest is parsed and passed to
+// stacktrace requests handlers.
+func TestLaunchRequestWithStackTraceDepth(t *testing.T) {
+	runTest(t, "increment", func(client *daptest.Client, fixture protest.Fixture) {
+		var stResp *dap.StackTraceResponse
+		runDebugSessionWithBPs(t, client,
+			// Launch
+			func() {
+				client.LaunchRequestWithArgs(map[string]interface{}{
+					"mode": "exec", "program": fixture.Path, "stackTraceDepth": 1,
+				})
+			},
+			// Set breakpoints
+			fixture.Source, []int{8},
+			[]func(){
+				// Stop at line 8
+				func() {
+					t.Helper()
+					client.StackTraceRequest(1, 0, 0)
+					stResp = client.ExpectStackTraceResponse(t)
+					expectFrames(t, stResp, 1000, 2, 2, 8)
+				}})
 	})
 }
 
-// runDebugSesion is a helper for executing the standard init and shutdown
+// runDebugSesionWithBPs is a helper for executing the standard init and shutdown
 // sequences for a program that does not stop on entry
-// while specifying unique launch criteria via parameters.
-func runDebugSession(t *testing.T, client *daptest.Client, launchRequest func()) {
+// while specifying breakpoints and unique launch criteria via parameters.
+func runDebugSessionWithBPs(t *testing.T, client *daptest.Client, launchRequest func(), source string, breakpoints []int, onBreakpoints []func()) {
```

Comment:
Please document the meaning of the `launchRequest` and `onBreakpoints` args. What is the called function(s) expected to do?

Reply:
Done